### ISSUE TITLE
Extra blocking via PM

### ISSUE THREAD

**Issue by HanabishiRecca** [MuntashirAkon/AppManager#205]
It will be good to add true blocking mode with root. Which actually **disables** activities/services/intents in the system.
Like it's done in MyAndroidTools, LuckyPatcher etc. (After using this tools such components appear in AppManager as disabled.)

**Comment by MuntashirAkon** [comment]
> It will be good to add true blocking mode with root.

Assuming you're talking about `pm` or Shizuku mode, Intent Firewall is actually far more better than the one you're proposing. See [FAQ](https://muntashirakon.github.io/AppManager/faq/app-components.html#why-are-the-components-blocked-by-am-not-detected-by-other-related-apps).

**Comment by HanabishiRecca** [comment]
Thanks for clarifying 😅

**Comment by HanabishiRecca** [comment]
@MuntashirAkon hello again. That's the thing why I asked about blocking. Now I collected some data. Seems like IFW not always can block an intent.

 For example, I'm blocking this:

![am](https://user-images.githubusercontent.com/13597663/103464836-a8d62780-4d58-11eb-9fbd-eabad233b4d2.png)

The app still shows ads:

![l2sd-ad](https://user-images.githubusercontent.com/13597663/103464852-c3a89c00-4d58-11eb-89d3-8e9908ae1588.png)

But when I'm disabling this intent via e.g. MAT (pm method I suppose):

![mat](https://user-images.githubusercontent.com/13597663/103464902-27cb6000-4d59-11eb-91fc-2fcf42589613.png)

Intent doesn't work and ad disappears:

![l2sd-noad](https://user-images.githubusercontent.com/13597663/103464913-387bd600-4d59-11eb-82dd-e32f494dfd52.png)

Am I get something wrong? Can you test this too?

**Comment by MuntashirAkon** [comment]
> Am I get something wrong? Can you test this too?

I can't seem to reproduce it. After you apply blocking via IFW, you may have to restart the application or invalidate cache. When you block using pm method, this is done by default. The ad might appear because it was stored in the cache or memory.

**Comment by HanabishiRecca** [comment]
Tried to even reboot the whole device. Also when I enabling intent back in MAT, ad is also immediately turns back. Regardless it's still blocked in AM.

**Comment by MuntashirAkon** [comment]
> Also when I enabling intent back in MAT, ad is also immediately turns back. Regardless it's still blocked in AM.

That not an issue as described in the FAQ.

> Tried to even reboot the whole device.

Check if the blocking is really applied by visiting `/data/system/ifw/<your package>.xml`

You've also failed to mention the AM version you're currently using.

**Comment by HanabishiRecca** [comment]
> That not an issue as described in the FAQ.
I mean AM block doesn't work even after pm block. 

> Check if the blocking is really applied by visiting /data/system/ifw/<your package>.xml
> You've also failed to mention the AM version you're currently using.

Sure. Latest AM Pre-release 2.5.21.
`/data/system/ifw/com.buak.Link2SD.xml` file is present and have a such content:
```xml
<rules>
<activity block="true" log="false">
  <component-filter name="com.buak.Link2SD/com.google.android.gms.ads.AdActivity"/>
</activity>
</rules>
```

Btw there are no any mentions of `IntentFirewall` or `ifw_intent_matched` events in the logcat. So I suppose this rules doesn't work for some reason. And I found a clue [here](https://carteryagemann.com/pages/android-intent-firewall.html):
> This directory is often `/data/system/ifw/`, **but not always**. 
> you can find the correct directory using the `getSystemSecureDirectory()` method

But in your app as I can see the path is hardcoded:
https://github.com/MuntashirAkon/AppManager/blob/77e18a38abe5b8fb74861078e790a3c298b9610c/app/src/main/java/io/github/muntashirakon/AppManager/rules/compontents/ComponentsBlocker.java#L62

Maybe my firmware just has the different IFW folder path?

**Comment by HanabishiRecca** [comment]
Hmm, seems like no. Some rules are definitely working. I can block app main activity and it won't launch.
Also I enabled IFW logging and confirmed that
```
I ifw_intent_matched: [0,com.buak.Link2SD/.Link2SD,10109,1,NULL,android.intent.action.MAIN,NULL,NULL,270532608]
```
But no matches for ads intent. Very weird 🙃

---
Fixed path maybe still need to be replaced by the function? Just as an compatibility improvement feature. I can create a new issue for that.

**Comment by MuntashirAkon** [comment]
> Fixed path maybe still need to be replaced by the function?

Documentation on Android source suggests that either `/data/system/ifw` (currently used) or `/data/secure/system/ifw` (used up to Android M) path can be used.

What's the Android version that you're using?

> But no matches for ads intent. Very weird 🙃

There could be other unknown factors at play here. But I can't seem to reproduce it on my device. Advertising doesn't seem to be working at all. Can you try an app called **Call Meter 3G** so that we can both be on the same page. If the findings are indeed correct, then it will be a very good finding.

**Comment by HanabishiRecca** [comment]
> What's the Android version that you're using?

10 (Q)

> Can you try an app called Call Meter 3G so that we can both be on the same page.

Ok. Blocking the ad here also doesn't work.

![cm](https://user-images.githubusercontent.com/13597663/103475251-6ac71b00-4dcd-11eb-8a94-e1d00be67de6.png)

![cm1](https://user-images.githubusercontent.com/13597663/103475260-7b779100-4dcd-11eb-9e25-18a7b0d46454.png)

**Comment by MuntashirAkon** [comment]
> Ok. Blocking the ad here also doesn't work.

There must be some reason behind this (my suspect is IPC). Anyway, I think it should be safe to conclude that it's better to disable these activities using _both_ IFW and PackageManager at the same time. So they will remain blocked in case the app enables them. Another idea would be to run a job service now and then to ensure that the components blocked in AM are actually blocked.

**Comment by HanabishiRecca** [comment]
> Anyway, I think it's safe to conclude that it's better to disable these components using both IFW and PackageManager at the same time.

Yes. Really not much of existing apps re-enabling intents. Also being visually in sync with other blocking tools is good too.

**Comment by MuntashirAkon** [comment]
It's definitely something to do with IPC. I've set SELinux to permissive and IFW isn't working even after I've set it to enforcing again. I'll file a report in the Google Bug Tracker (although there's hardly any chance that this issue will be addressed). They've also neglected my other issue on app ops. They only like the easy-to-fix bugs.

**Comment by MuntashirAkon** [comment]
> Really not much of existing apps re-enabling intents.

People usually blocks components from the dodgy apps, not just from the _innocent_-looking apps containing AdActivity.

Anyway, thank you for your detailed investigations on this matter. This issue is now closed via 4732500f and 58ca8051. Job service is related to #61 and will be addressed there.

**Comment by HanabishiRecca** [comment]
> Anyway, thank you for your detailed investigations on this matter. This issue is now closed via 4732500 and 58ca805.

My respect, sir!
